### ISSUE TITLE
Add option to pre-babelify for production

### ISSUE THREAD

**Issue by gigabo** [redfin/react-server#773]
Currently the CLI always runs through babel register to run transforms on the fly.  Need an option to save the transformed es5 output and run straight from that in production for faster startup.

**Comment by gigabo** [comment]
I guess another option here would be to run webpack bundles on the server.  I remember thinking at one time that there were some benefits to that, but I can't recall what they were at the moment. 😅

**Comment by doug-wade** [comment]
@gigabo the reason that I remember off the top of my head is making the server and the client build easier to reason about.  Currently, you have to ship everything twice, once as a `require` hook server-side, and once as a webpack hook client-side.  If we moved to webpack-only, it would likely be easier to reason about and explain (and would give users the ability to modify how modules are loaded server-side, which they can't do currently, by modifying their webpack config).

**Comment by aickin** [comment]
@gigabo are you talking about the use of `babel-register` to transform files as part of `require` on the server side? If so, I'm pretty sure that it [keeps a cache in between runs](https://babeljs.io/docs/usage/babel-register/#environment-variables) automagically. 

Now, maybe it's not using the cache for some reason, but if it is, I'd guess that pre-transforming isn't going to help very much.

**Comment by drewpc** [comment]
Maybe this is a bit off topic, but I'm for anything that makes the client and server sides mirror each other.  If Webpack is used for the client side, it should be for the server side as well.  It seems easier to understand/develop a truly isomorphic app if the client and server are as close to the same as possible.

**Comment by gigabo** [comment]
@aickin Oh, interesting.  Maybe we could set `process.env.BABEL_CACHE_PATH` to the build directory and ship that (along with the webpack manifest.json) for fast startup in production?

@drewpc Not off topic at all!  That's a good point about keeping the client and server as close to each other as possible.  I'm not so familiar with webpack on the server.  Is that a common practice?

**Comment by roblg** [comment]
I think next.js is using webpack on the server as well. One benefit is that then you can just delete the single built asset from the `require.cache`, and you get server-side "hot reload". (Or if you have multiple bundles, you have to delete more than one, but still)

**Comment by gigabo** [comment]
> server-side "hot reload"

![ib](https://cloud.githubusercontent.com/assets/115908/20806827/45442de8-b7b1-11e6-9ee8-1d0f0a0c1b46.gif)

**Comment by drewpc** [comment]
@gigabo I'm new to Webpack as a whole--not an expert!  From what I've read and messed around with in planning my current project, you can generate Webpack chunks for the browser and for use in node in a server environment.  It would be pre-compiled, pre-babelified, etc.  Here's some links I was looking at a while back:

[https://webpack.github.io/docs/configuration.html#target](https://webpack.github.io/docs/configuration.html#target)
[https://www.npmjs.com/package/universal-webpack](https://www.npmjs.com/package/universal-webpack)
[http://jlongster.com/Backend-Apps-with-Webpack--Part-I](http://jlongster.com/Backend-Apps-with-Webpack--Part-I)

Essentially, I see moving to Webpack on the server solves three problems:
1.  Consistency between client/server (i.e. "requiring" SCSS/images)
2.  Handles pre-babelify case
3.  Facilitates server-side HMR

**Comment by gigabo** [comment]
Awesome.  Thanks for the links and the summary of benefits @drewpc!

Anyone know of any _drawbacks_ to webpack for the server build?

**Comment by drewpc** [comment]
I can think of two that I'm considering/researching right now:

1.  It *seems* like Webpack will output two separate sets of libraries (one for client and one for server) depending on what the output target is.  That could lead to a duplication of assets (like images, CSS, etc).  I'm looking into what people do to address that now.
2.  There might be a performance penalty of using babelified code in a modern version of Node.  This is just a theory--no data to back it up.  I can imagine someone writing a kick ass app in ES6 or ES7, running it on the hottest versions of Node and React and then Webpack transpiles the code down to (essentially) a lowest-common-denominator version.  Having said that, maybe that would be a reason to keep Webpack outputting two sets of libraries.  We could keep the server side at a more recent Javascript version and dumb down the browser side.

Sorry...my fingers are typing the stream of consciousness as I think this through...

**Comment by gigabo** [comment]
I think we suffer from that second problem already.  We're running everything through babel with a fixed configuration.  We could use [`babel-preset-env`](https://github.com/babel/babel-preset-env) to address that, but AFAICT it's not affected by whether or not we're using webpack on the server.

The first one is interesting... :thinking:

**Comment by aickin** [comment]
@drewpc I agree with you totally about the benefits of server webpack, but there are some tradeoffs (or at least there were last time I looked at this). The biggest issue for me with server side webpack is timing of the debug-test cycle and debugging/sourcemaps experience. A lot has changed since we went with the server not using webpack, though, so the downsides have probably changed. 

Certainly a good thing to investigate!

**Comment by doug-wade** [comment]
@drewpc surprisingly, using the newest js features in node often [leads to worse performance, rather than better](https://kpdecker.github.io/six-speed/).

**Comment by aickin** [comment]
Big +1 to what @doug-wade said, though browsers are in the process of optimizing ES2015+, so the landscape is ever changing. My total shot-in-the-dark guess is that the balance will be in favor of native impls in v8 by maybe mid-next year and hopefully in Node by late 2017. We'll see.

There can be a significant perf benefit **in browsers** to not transforming code because native ES2015+ code is often fewer bytes than Babel-transformed code, which makes it smaller over the wire and faster during parse. But in node, the wire size is moot, and the parse benefit is, I think, a pretty tiny sliver of the overall server time.

**Comment by jhnns** [comment]
Hi, just chiming in as a webpack contributor - I have been looking at your project and I'm quite impressed by what you've achieved with universal rendering, especially the streaming!

When trying to integrate this in a project, I noticed that the server code was not being webpacked, which caused a bit of confusion.

With the current setup this uses two different module runtimes: one that allows asynchronous transforms via loaders (your client bundle) and one that doesn't (your node code).

If the server code was webpacked, this would open up opportunities such as:

- importing images into React modules (like `import someImage from './someImage.jpg`)
- using CSS modules as in #689 (because the hashed class names are needed in the markup)
- setting split points via `require.ensure` (webpack 1) or [dynamic `import()`](https://github.com/tc39/proposal-dynamic-import)) (webpack 2)

I would be happy to shed some light on integration methods if this is something you're interested in.

What were the original reasons to not use webpack for server-side code?

**Comment by drewpc** [comment]
@jhnns I'm working on adding this capability right now and would love your input when I submit the PR in a few days.
